cpp/CMakeLists.txt

```diff
@@ -690,7 +690,9 @@ if (ARROW_JEMALLOC)
   add_definitions(-DARROW_JEMALLOC)
   add_definitions(-DARROW_JEMALLOC_INCLUDE_DIR=${JEMALLOC_INCLUDE_DIR})
 
-  if (CMAKE_COMPILER_IS_GNUCXX AND PTHREAD_LIBRARY)
+  if ((CMAKE_COMPILER_IS_GNUCXX OR
+        (CMAKE_CXX_COMPILER_ID MATCHES "Clang"))
+      AND PTHREAD_LIBRARY)
```

Comment:
I found that linking pthread is also necessary for me when using clang-6 on Linux with Ubuntu 14.04